File: platform/macos/Bridge/NativeListener.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Combine
import Foundation
import TidyVNC

extension tidyvnc_listener_options: ABIValue {}
extension tidyvnc_listener_snapshot: ABIValue {}
extension tidyvnc_listener_event: ABIValue {}

public struct NativeListenOptions: Sendable {
  public var address = ""
  public var port: UInt32 = 5500
  public var ipv4 = true, ipv6 = true
  public var backlog: UInt32 = 16, pendingCapacity: UInt32 = 8, eventCapacity: UInt32 = 32
  public var pendingTimeoutMilliseconds: UInt32 = 30000
  public init() {}
}
public struct NativeListenerAddress: Sendable, Equatable {
  public let host: String
  public let port: UInt32
  init(_ value: tidyvnc_listener_address) {
    host = withUnsafeBytes(of:value.host) { String(decoding:$0.prefix(while: { $0 != 0 }),as:UTF8.self) }
    port = value.port
  }
}
public struct NativeIncomingPeer: Sendable, Equatable, Identifiable {
  public let id: UInt64
  public let address: NativeListenerAddress
  fileprivate let listener: UInt64
}
public enum NativeListenerState: UInt32, Sendable { case starting, listening, stopping, closed, failed }
public enum NativeListenerFailure: UInt32, Sendable { case none, cancelled, bind, accept, invalidAddress, unsupported, eventOverflow, internalFailure }
public struct NativeListenerSnapshot: Sendable, Equatable {
  public let state: NativeListenerState
  public let failure: NativeListenerFailure
  public let nativeError: Int32
  public let pending: UInt32
  public let addresses: [NativeListenerAddress]
  init(_ value: tidyvnc_listener_snapshot) {
    state = NativeListenerState(rawValue:value.state) ?? .failed
    failure = NativeListenerFailure(rawValue:value.error) ?? .internalFailure
    nativeError = value.native_error; pending = value.pending
    addresses = [value.addresses.0,value.addresses.1].prefix(Int(min(value.address_count,2))).map(NativeListenerAddress.init)
  }
}

// One listener owns its event stream. Readiness is coalesced by the C dispatcher
// and NativeDelivery; no recurring state polling or per-peer observer is needed.
@MainActor public final class NativeListener: ObservableObject {
  @Published public private(set) var snapshot: NativeListenerSnapshot
  @Published public private(set) var incoming: [NativeIncomingPeer] = []
  @Published public private(set) var deliveryError: NativeError?
  public private(set) var isClosing = false
  private let runtime: NativeRuntime
  private let handle: NativeHandle
  private var subscription: NativeHandle?
  private var delivery: NativeDelivery?
  private var closeTask: Task<Void, Error>?
  init(runtime: NativeRuntime, options: NativeListenOptions) throws {
    self.runtime = runtime
    var config = abi(tidyvnc_listener_options.self)
    try checked { tidyvnc_listener_options_init(&config,$0) }
    config.port = options.port; config.ipv4 = options.ipv4 ? 1 : 0; config.ipv6 = options.ipv6 ? 1 : 0
    config.backlog = options.backlog; config.pending_capacity = options.pendingCapacity; config.event_capacity = options.eventCapacity
    config.pending_timeout_ms = options.pendingTimeoutMilliseconds
    var raw: UInt64 = 0
    try withText(options.address) { address in
      config.address = address; try checked { tidyvnc_listener_create(runtime.handle.raw,&config,&raw,$0) }
    }
    let owner = NativeHandle(adopting:raw); handle = owner
    var initial = abi(tidyvnc_listener_snapshot.self)
    try checked { tidyvnc_listener_get_snapshot(raw,&initial,$0) }
    snapshot = NativeListenerSnapshot(initial)
    let context = NativeDelivery { [weak self] _,_ in self?.receive() }
    delivery = context
    var callbacks = abi(tidyvnc_callbacks.self)
    callbacks.context = Unmanaged.passUnretained(context).toOpaque()
    callbacks.retain_context = retainNativeDelivery; callbacks.release_context = releaseNativeDelivery; callbacks.ready = readyNativeDelivery
    var subscribed: UInt64 = 0
    _ = try withExtendedLifetime(context) { try checked { tidyvnc_listener_subscribe(owner.raw,&callbacks,&subscribed,$0) } }
    subscription = NativeHandle(adopting:subscribed)
    // Consume queued history before exposing this owner. Otherwise the current
    // snapshot above could be Listening while the first queued event is Starting.
    receive()
  }
  private func receive() {
    guard !isClosing else { return }
    do {
      var event = abi(tidyvnc_listener_event.self)
      while try checked(allowing:[.ok,.noChange],{ tidyvnc_listener_take_event(handle.raw,&event,$0) }) == .ok {
        snapshot = NativeListenerSnapshot(event.snapshot)
        if event.kind == UInt32(TIDYVNC_LISTENER_INCOMING) {
          incoming.append(.init(id:event.incoming_id,address:NativeListenerAddress(event.peer),listener:handle.raw))
        } else if event.incoming_id != 0 {
          incoming.removeAll { $0.id == event.incoming_id }
        }
        if [.stopping,.closed,.failed].contains(snapshot.state) { incoming.removeAll() }
      }
    } catch { deliveryError = error as? NativeError ?? NativeError(.internalFailure,"Listener delivery failed") }
  }
  private func validate(_ peer: NativeIncomingPeer) throws {
    guard !isClosing else { throw NativeError(.closing,"Listener is closing") }
    guard peer.listener == handle.raw, incoming.contains(peer) else { throw NativeError(.stale,"Incoming connection is no longer available") }
  }
  public func reject(_ peer: NativeIncomingPeer) throws {
    try validate(peer); defer { receive() }
    try checked { tidyvnc_listener_reject(handle.raw,peer.id,$0) }
  }
  public func accept(_ peer: NativeIncomingPeer, into session: NativeSession) async throws -> NativeCompletion {
    try validate(peer)
    return try await session.acceptIncoming(listener:handle,incoming:peer.id)
  }
  // Stop preserves terminal event delivery and does not close accepted sessions.
  public func stop() throws { try checked { tidyvnc_listener_stop(handle.raw,$0) } }
  func beginClose() -> Task<Void,Error> {
    if let closeTask { return closeTask }
    isClosing = true; incoming.removeAll()
    let owner = handle, subscribed = subscription, context = delivery
    context?.invalidate()
    if let subscribed { _ = tidyvnc_subscription_unsubscribe(subscribed.raw,nil) }
    _ = tidyvnc_listener_stop(owner.raw,nil)
    let task = Task {
      var failure: (any Error)?
      do { try await waitForNativeDrain(owner,.listener) } catch { failure = error }
      if let subscribed {
        do { try await waitForNativeDrain(subscribed,.subscription) } catch { if failure == nil { failure = error } }
      }
      await context?.drain()
      if let failure { throw failure }
    }
    closeTask = task; return task
  }
  public func close() async throws { try await beginClose().value }
  deinit {
    delivery?.invalidate()
    if let subscription { _ = tidyvnc_subscription_unsubscribe(subscription.raw,nil) }
    _ = tidyvnc_listener_stop(handle.raw,nil)
  }
}

File: platform/windows/TidyVNC.Native/Bridge/NativeListener.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using System.Collections.ObjectModel;
using CommunityToolkit.Mvvm.ComponentModel;
using TidyVNC.Native.Interop;

namespace TidyVNC.Native;

public sealed class NativeListenOptions
{
    public string Address { get; set; } = "";
    public uint Port { get; set; } = 5500;
    public bool Ipv4 { get; set; } = true;
    public bool Ipv6 { get; set; } = true;
    public uint Backlog { get; set; } = 16;
    public uint PendingCapacity { get; set; } = 8;
    public uint EventCapacity { get; set; } = 32;
    public uint PendingTimeoutMilliseconds { get; set; } = 30_000;
}

public readonly record struct NativeListenerAddress(string Host, uint Port)
{
    internal static unsafe NativeListenerAddress From(tidyvnc_listener_address value)
        => new(NativeText.Fixed(value.host, 64), value.port);

    /// <summary>True for addresses reachable only from this computer (no firewall prompt).</summary>
    public bool IsLoopback => Host is "127.0.0.1" or "::1" || Host.StartsWith("127.", StringComparison.Ordinal);
}

public sealed record NativeIncomingPeer(ulong Id, NativeListenerAddress Address)
{
    internal ulong Listener { get; init; }
}

public enum NativeListenerState : uint { Starting, Listening, Stopping, Closed, Failed }
public enum NativeListenerFailure : uint { None, Cancelled, Bind, Accept, InvalidAddress, Unsupported, EventOverflow, InternalFailure }

public sealed record NativeListenerSnapshot(NativeListenerState State, NativeListenerFailure Failure, int NativeError, uint Pending,
                                            IReadOnlyList<NativeListenerAddress> Addresses)
{
    internal static NativeListenerSnapshot From(tidyvnc_listener_snapshot value)
    {
        var addresses = new List<NativeListenerAddress>();
        for (var i = 0; i < Math.Min(value.address_count, 2u); i++) addresses.Add(NativeListenerAddress.From(value.addresses[i]));
        return new(Enum.IsDefined((NativeListenerState)value.state) ? (NativeListenerState)value.state : NativeListenerState.Failed,
                   Enum.IsDefined((NativeListenerFailure)value.error) ? (NativeListenerFailure)value.error : NativeListenerFailure.InternalFailure,
                   value.native_error, value.pending, addresses);
    }

    public bool Equals(NativeListenerSnapshot? other)
        => other is not null && State == other.State && Failure == other.Failure && NativeError == other.NativeError &&
           Pending == other.Pending && Addresses.SequenceEqual(other.Addresses);

    public override int GetHashCode() => HashCode.Combine(State, Failure, NativeError, Pending, Addresses.Count);
}

/// <summary>
/// Reverse-connection listener (NativeListener.swift). One owner of its event
/// stream; readiness is coalesced like sessions, with no state polling.
/// </summary>
public sealed partial class NativeListener : ObservableObject
{
    private readonly NativeRuntime runtime;
    private readonly NativeHandle handle;
    private readonly NativeHandle? subscription;
    private readonly NativeDelivery? delivery;
    private Task? closeTask;

    [ObservableProperty] public partial NativeListenerSnapshot Snapshot { get; private set; }
    [ObservableProperty] public partial NativeError? DeliveryError { get; private set; }
    public ObservableCollection<NativeIncomingPeer> Incoming { get; } = [];
    public bool IsClosing { get; private set; }

    internal unsafe NativeListener(NativeRuntime runtime, NativeListenOptions options)
    {
        this.runtime = runtime;
        var config = Abi.Init<tidyvnc_listener_options>();
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_listener_options_init(&config, &error), &error);
        config.port = options.Port; config.ipv4 = options.Ipv4 ? 1u : 0u; config.ipv6 = options.Ipv6 ? 1u : 0u;
        config.backlog = options.Backlog; config.pending_capacity = options.PendingCapacity;
        config.event_capacity = options.EventCapacity; config.pending_timeout_ms = options.PendingTimeoutMilliseconds;
        var address = NativeText.Utf8(options.Address);
        ulong raw = 0;
        fixed (byte* p = address)
        {
            config.address = NativeText.Span(p, address.Length);
            Abi.Check(NativeMethods.tidyvnc_listener_create(runtime.Handle.Raw, &config, &raw, &error), &error);
        }
        handle = NativeHandle.Adopt(raw);
        var initial = Abi.Init<tidyvnc_listener_snapshot>();
        Abi.Check(NativeMethods.tidyvnc_listener_get_snapshot(raw, &initial, &error), &error);
        Snapshot = NativeListenerSnapshot.From(initial);
        var weak = new WeakReference<NativeListener>(this);
        delivery = new NativeDelivery(runtime.Dispatcher, (_, _) => { if (weak.TryGetTarget(out var listener)) listener.Receive(); });
        var callbacks = delivery.Callbacks();
        ulong subscribed = 0;
        if (NativeMethods.tidyvnc_listener_subscribe(handle.Raw, &callbacks, &subscribed, &error) != Tidyvnc.TIDYVNC_OK)
        {
            delivery.AbandonUnretained();
            _ = NativeMethods.tidyvnc_listener_stop(handle.Raw, null);
            handle.Dispose();
            throw new NativeError(error);
        }
        subscription = NativeHandle.Adopt(subscribed);
        // Consume queued history first, so the first queued Starting event
        // cannot follow an already-Listening snapshot.
        Receive();
    }

    private unsafe void Receive()
    {
        if (IsClosing) return;
        try
        {
            var value = Abi.Init<tidyvnc_listener_event>();
            var error = Abi.Init<tidyvnc_error>();
            while (Abi.Check(NativeMethods.tidyvnc_listener_take_event(handle.Raw, &value, &error), &error, NativeStatus.Ok, NativeStatus.NoChange) == NativeStatus.Ok)
            {
                Snapshot = NativeListenerSnapshot.From(value.snapshot);
                if (value.kind == Tidyvnc.TIDYVNC_LISTENER_INCOMING)
                    Incoming.Add(new NativeIncomingPeer(value.incoming_id, NativeListenerAddress.From(value.peer)) { Listener = handle.Raw });
                else if (value.incoming_id != 0)
                    Remove(value.incoming_id);
                if (Snapshot.State is NativeListenerState.Stopping or NativeListenerState.Closed or NativeListenerState.Failed) Incoming.Clear();
            }
        }
        catch (NativeError problem) { DeliveryError = problem; }
    }

    private void Remove(ulong id)
    {
        for (var i = Incoming.Count - 1; i >= 0; i--) if (Incoming[i].Id == id) Incoming.RemoveAt(i);
    }

    private void Validate(NativeIncomingPeer peer)
    {
        UiThread.Require(runtime.Dispatcher);
        if (IsClosing) throw new NativeError(NativeStatus.Closing, "Listener is closing");
        if (peer.Listener != handle.Raw || !Incoming.Contains(peer))
            throw new NativeError(NativeStatus.Stale, "Incoming connection is no longer available");
    }

    public unsafe void Reject(NativeIncomingPeer peer)
    {
        Validate(peer);
        try
        {
            var error = Abi.Init<tidyvnc_error>();
            Abi.Check(NativeMethods.tidyvnc_listener_reject(handle.Raw, peer.Id, &error), &error);
        }
        finally { Receive(); }
    }

    public Task<NativeCompletion> AcceptAsync(NativeIncomingPeer peer, NativeSession session)
    {
        Validate(peer);
        return session.AcceptIncomingAsync(handle, peer.Id);
    }

    /// <summary>Stops listening; terminal events still arrive and accepted sessions stay open.</summary>
    public unsafe void Stop()
    {
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_listener_stop(handle.Raw, &error), &error);
    }

    internal Task BeginClose()
    {
        if (closeTask is not null) return closeTask;
        IsClosing = true;
        Incoming.Clear();
        delivery?.Invalidate();
        unsafe
        {
            if (subscription is not null) _ = NativeMethods.tidyvnc_subscription_unsubscribe(subscription.Raw, null);
            _ = NativeMethods.tidyvnc_listener_stop(handle.Raw, null);
        }
        closeTask = Close();
        return closeTask;
    }

    private async Task Close()
    {
        Exception? failure = null;
        try { await NativeDrainer.WaitAsync(handle, NativeDrain.Listener).ConfigureAwait(true); } catch (Exception e) { failure = e; }
        if (subscription is not null)
        {
            try { await NativeDrainer.WaitAsync(subscription, NativeDrain.Subscription).ConfigureAwait(true); }
            catch (Exception e) { failure ??= e; }
        }
        if (delivery is not null) await delivery.DrainAsync().ConfigureAwait(true);
        if (failure is not null) throw failure;
    }

    public Task CloseAsync()
    {
        UiThread.Require(runtime.Dispatcher);
        return BeginClose();
    }

    ~NativeListener()
    {
        delivery?.Invalidate();
        unsafe
        {
            if (subscription is { IsClosed: false }) _ = NativeMethods.tidyvnc_subscription_unsubscribe(subscription.Raw, null);
            if (!handle.IsClosed) _ = NativeMethods.tidyvnc_listener_stop(handle.Raw, null);
        }
    }
}
